src/document.c

```diff
@@ -930,36 +930,89 @@ typedef struct
 } FileData;
 
 
+static time_t get_mtime(const gchar *locale_filename)
+{
+	GError *error = NULL;
+	const gchar *err_msg = NULL;
+	time_t time = 0;
+
+	if (!file_prefs.use_safe_file_saving && file_prefs.use_gio_unsafe_file_saving)
+	{
+		GFile *file = g_file_new_for_path(locale_filename);
+		GFileInfo *info = g_file_query_info(file, G_FILE_ATTRIBUTE_TIME_MODIFIED, G_FILE_QUERY_INFO_NONE, NULL, &error);
+
+		if (info)
+		{
+			GTimeVal timeval;
+
+			g_file_info_get_modification_time(info, &timeval);
+			g_object_unref(info);
+			time = timeval.tv_sec;
+		}
+		else if (error)
+			err_msg = error->message;
+
+		g_object_unref(file);
+	}
+	else
+	{
+		struct stat st;
+
+		if (g_stat(locale_filename, &st) == 0)
+			time = st.st_mtime;
+		else
+			err_msg = g_strerror(errno);
+	}
+
+	if (err_msg)
+	{
+		gchar *utf8_filename = utils_get_utf8_from_locale(locale_filename);
+
+		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
+			utf8_filename, err_msg);
+		g_free(utf8_filename);
+	}
+
+	if (error)
+		g_error_free(error);
+
+	return time;
+}
+
+
 /* loads textfile data, verifies and converts to forced_enc or UTF-8. Also handles BOM. */
 static gboolean load_text_file(const gchar *locale_filename, const gchar *display_filename,
 	FileData *filedata, const gchar *forced_enc)
 {
 	GError *err = NULL;
-	struct stat st;
 
 	filedata->data = NULL;
 	filedata->len = 0;
 	filedata->enc = NULL;
 	filedata->bom = FALSE;
 	filedata->readonly = FALSE;
+	filedata->mtime = get_mtime(locale_filename);
 
-	if (g_stat(locale_filename, &st) != 0)
-	{
-		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
-			display_filename, g_strerror(errno));
+	if (filedata->mtime == 0)
 		return FALSE;
-	}
 
-	filedata->mtime = st.st_mtime;
+	if (!file_prefs.use_safe_file_saving && file_prefs.use_gio_unsafe_file_saving)
+	{
+		GFile *file = g_file_new_for_path(locale_filename);
+
+		g_file_load_contents(file, NULL, &filedata->data, &filedata->len, NULL, &err);
+		g_object_unref(file);
+	}
+	else
+		g_file_get_contents(locale_filename, &filedata->data, &filedata->len, &err);
 
-	if (! g_file_get_contents(locale_filename, &filedata->data, NULL, &err))
+	if (err)
 	{
 		ui_set_statusbar(TRUE, "%s", err->message);
 		g_error_free(err);
 		return FALSE;
 	}
 
-	filedata->len = (gsize) st.st_size;
```

Comment:
@techee as @elextr said, you don't pass the `strlen()` value (so, truncated at the first `\0`), but the data size as read by `g_file_*_contents()` (which should include possible embedded `\0`s).  So it should effectively be the same as the `stat()` size, or if it's not anyway it would lead to an out-of-bounds read (which probably could happen currently if a race condition happens between the 2 calls and the file changes size in-between).

So no, I don't think you need to change this, as AFAIU, the stat value was never really documented as a better alternative to the `g_file_get_contents()` one (see my tracing back) and has glaring potential drawbacks (invalid memory access).

> Maybe we shouldn't store the length here at all if encoding stores strlen() as part of its NUL checks.

Yes we need it, as encodings checks need the real length value, although currently it doesn't like it when it differs from `strlen()`, but has special handling especially for this case.